src/h3lib/lib/localij.c

```diff
@@ -542,3 +540,109 @@ int H3_EXPORT(h3Distance)(H3Index origin, H3Index h3) {
 
     return ijkDistance(&originIjk, &h3Ijk);
 }
+
+/**
+ * Number of indexes in a line from the start index to the end index,
+ * to be used for allocating memory. Returns a negative number if the
+ * line cannot be computed.
+ *
+ * @param start Start index of the line
+ * @param end End index of the line
+ * @return Size of the line, or a negative number if the line cannot
+ * be computed.
+ */
+int H3_EXPORT(h3LineSize)(H3Index start, H3Index end) {
+    int distance = H3_EXPORT(h3Distance)(start, end);
+    return distance >= 0 ? distance + 1 : distance;
+}
+
+/**
+ * Given cube coords as floats, round to valid integer coordinates. Algorithm
+ * from https://www.redblobgames.com/grids/hexagons/#rounding
+ * @param i   Floating-point I coord
+ * @param j   Floating-point J coord
+ * @param k   Floating-point K coord
+ * @param ijk IJK coord struct, modified in place
+ */
+static void cubeRound(float i, float j, float k, CoordIJK* ijk) {
+    int ri = round(i);
+    int rj = round(j);
+    int rk = round(k);
+
+    float iDiff = fabs((float)ri - i);
+    float jDiff = fabs((float)rj - j);
+    float kDiff = fabs((float)rk - k);
+
+    // Round, maintaining valid cube coords
+    if (iDiff > jDiff && iDiff > kDiff) {
+        ri = -rj - rk;
+    } else if (jDiff > kDiff) {
+        rj = -ri - rk;
+    } else {
+        rk = -ri - rj;
+    }
+
+    ijk->i = ri;
+    ijk->j = rj;
+    ijk->k = rk;
+}
+
+/**
+ * Given two H3 indexes, return the line of indexes between them (inclusive).
+ *
+ * This function may fail to find the line between two indexes, for
+ * example if they are very far apart. It may also fail when finding
+ * distances for indexes on opposite sides of a pentagon.
+ *
+ * Notes:
+ *
+ *  - The specific output of this function should not be considered stable
+ *    across library versions. The only guarantees the library provides are
+ *    that the line length will be `h3Distance(start, end) + 1` and that
+ *    every index in the line will be a neighbor of the preceding index.
+ *  - Lines are drawn in grid space, and may not correspond exactly to either
+ *    Cartesian lines or great arcs.
+ *
+ * @param start Start index of the line
+ * @param end End index of the line
+ * @param out Output array, which must be of size h3LineSize(start, end)
+ * @return 0 on success, or another value on failure.
+ */
+int H3_EXPORT(h3Line)(H3Index start, H3Index end, H3Index* out) {
+    int distance = H3_EXPORT(h3Distance)(start, end);
```

Comment:
Unfortunate that this effectively has to be called twice (once for the mem allocation and once for the safeguard here).

Perhaps there can be two functions, here, one that takes the chunk of memory and it's size and one with this signature that takes just the chunk of memory, assumes that it's big enough, and calls `h3Distance` to determine the usable size (or fail out early).

This would allow h3Distance to conceivably be called only once by users wanting to squeeze as much performance out of these calls as they can (such as the potential `polyfill` algorithm to reduce memory consumption and maybe run quicker, too).

Reply:
Well, `h3Distance` is constant time and pretty fast, so I'm not sure this is really an issue. `polyfill`, `kRing`, etc all have the same problem. But if we care, then the simplest solution is to require the caller to pass in the size of the memory allocation:
```
H3_EXPORT(h3Line)(H3Index start, H3Index end, H3Index* out, int numIndexes);
```
All callers ought to call `h3LineSize` in any case, and can just pass the result in. The only cost is a slightly uglier API that's inconsistent with `kRing` and others.

Reply:
Well, if `h3Distance` doesn't add a significant amount of time, then we don't have to worry about it.